Scene-to-Image Prompt Layering System

xaedes opened the pull request:
# Summary of the change

- new Scene-to-Image tab
- new scn2img function
- functions for loading and running monocular_depth_estimation with tensorflow

# Description

(relevant motivation, which issue is fixed)

Related to discussion #925

>  Would it be possible to have a layers system where we could do have foreground, mid, and background objects which relate to one another and share the style? So we could say generate a landscape, one another layer generate a castle, and on another layer generate a crowd of people.

To make this work I made a prompt-based layering system in a new "Scene-to-Image" tab. 
You write a a multi-line prompt that looks like markdown, where each section declares one layer.
It is hierarchical, so each layer can have their own child layers.

Examples: https://imgur.com/a/eUxd5qn
![](https://i.imgur.com/L61w00Q.png)

In the frontend you can find a brief documentation for the syntax, examples and reference for the various arguments.

Here a short summary:

Sections with "prompt" and child layers are img2img, without child layers they are txt2img.
Without "prompt" they are just images, useful for mask selection, image composition, etc.
Images can be initialized with "color", resized with "resize" and their position specified with "pos". 
Rotation and rotation center are "rotation" and "center". 

Mask can automatically be selected by color or by estimated depth based on https://huggingface.co/spaces/atsantiago/Monocular_Depth_Filter.

![](https://i.imgur.com/8rMHWmZ.png)

# Additional dependencies that are required for this change

For mask selection by monocular depth estimation tensorflow is required and the model must be cloned to ./src/monocular_depth_estimation/ 
Changes in environment.yaml:
- einops>=0.3.0
- tensorflow>=2.10.0 

Einops must be allowed to be newer for tensorflow to work.

# Checklist:

- [x] I have changed the base branch to `dev`
- [x] I have performed a self-review of my own code
- [x] I have commented my code in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

Kasiya13 commented:
Holy shit this is awesome! I'll have to try it later 😯
Have you tried adding it to the streamlit UI in the dev branch?

Kasiya13 commented:
If you're not on the discord yet, feel free to join: https://discord.gg/gyXNe4NySY

Dekker3D commented:
This is interesting, and having a depth estimator built in would be nice for my own shenanigans too. I've been attempting to build a 3D scene based on generated images, going back and forth between rendering and generating. Is this depth estimator better than MiDaS? The depth output I see for MiDaS looks good enough to build a 3D model from, and someone has been doing just that!

- the post about someone making 3D assets using MiDaS and SD (NSFW): https://www.reddit.com/r/singularity/comments/xdjdno/locus_tenebrarum_an_nsfw_video_game_made_almost/
- A MiDaS demo: https://huggingface.co/spaces/pytorch/MiDaS

xaedes commented:
> Holy shit this is awesome! I'll have to try it later 😯 Have you tried adding it to the streamlit UI in the dev branch?

I have looked intensively into the streamlit code. 
Unfortunately the streamlit code is currently structured in a way that strongly couples it with the (streamlit) frontend. 
The only way to support it for both gradio and streamlit in a consistent way with the rest of the code seems to be code duplication. 

I suppose I could do that. 
But currently I feel it makes more sense to decouple the streamlit code...

xaedes commented:
> This is interesting, and having a depth estimator built in would be nice for my own shenanigans too. I've been attempting to build a 3D scene based on generated images, going back and forth between rendering and generating. Is this depth estimator better than MiDaS? The depth output I see for MiDaS looks good enough to build a 3D model from, and someone has been doing just that!
> 
> * the post about someone making 3D assets using MiDaS and SD (NSFW): https://www.reddit.com/r/singularity/comments/xdjdno/locus_tenebrarum_an_nsfw_video_game_made_almost/
> * A MiDaS demo: https://huggingface.co/spaces/pytorch/MiDaS

MiDaS looks nice, I will try it, other and/or better depth estimation is always welcome. Looks easy enough to use.

> build a 3D scene based on generated images, going back and forth between rendering and generating

That sounds nice, was also thinking that something like that should be possible.

Kasiya13 commented:
> > Holy shit this is awesome! I'll have to try it later 😯 Have you tried adding it to the streamlit UI in the dev branch?
> 
> I have looked intensively into the streamlit code. Unfortunately the streamlit code is currently structured in a way that strongly couples it with the (streamlit) frontend. The only way to support it for both gradio and streamlit in a consistent way with the rest of the code seems to be code duplication.
> 
> I suppose I could do that. But currently I feel it makes more sense to decouple the streamlit code...

Oh that's actually being worked on, rn the goal is to get it decoupled and have an API backend to let anyone integrate it with whatever UI they'd like :)

If you'd like to help, you're welcome to join the discord if you're not over there already.

codedealer reviewed (changes requested):
@xaedes thanks for the contribution! It's quite extensive one so I had to poke around a bit.

As far as I can see the basic workflow works fine there are only a couple notes about integration with the rest of the app:
"View batch progress" didn't work for me no matter what I tried. But that's not a big deal if there's no easy fix.
"Output info" is currently spews some debug information that is not really useful for the end user and can't be used to generate something as the label of the textbox suggest.

If you could look at that it'd be great, thanks!

xaedes commented:
Thank you for taking a look at it, I appreciate it!

Regarding the "View batch progress" I have no opinion right now, will have to look into that first.
Oh yea, you are totally right. The current output info is no good, I will improve on it!

codedealer commented:
@xaedes I went and checked the current `dev` and it looks like the output is broken in there for img2img too so if you can fix it for both (I assume you based yours on img2img) in this PR I'd appreciate it, if not I guess it will have to be fixed separately.

xaedes commented:
@codedealer batch progress for txt2img and img2img on dev works for me. it also works for txt2img and img2img on this PR branch. but not for scn2img as I don't handle job_info properly. I am working on that right now.

codedealer commented:
@xaedes yeah I wasn't clear on that, sorry. Batch progress works on txt2img/img2img for me as well. There are two separate issues currently
* Batch progress on scn2img
* Generated output info in the tab on img2img and scn2img — this one was not introduced by this PR but if you're going to look into it I wanted to ask that you fix both scn2img and img2img.

hlky commented on environment.yaml:
```diff
@@ -23,7 +23,7 @@ dependencies:
     - basicsr>=1.3.4.0
     - diffusers==0.3.0
-    - einops==0.3.0
+    - einops>=0.3.0
```
Why was this changed? If there's a new version and it does not cause issues, specify the new version, otherwise versions should be pinned to avoid breaking changes.

xaedes replied:
Tensorflow needs einops>=0.3.0 otherwise, "AttributeError: module 'keras.backend' has no attribute 'is_tensor'" is thrown when loading keras model. Tensorflow is for depth from image estimation, used for automatic mask selection in scn2img.

I changed it to pin the versions.

hlky commented on scripts/webui.py:
```diff
@@ -659,43 +771,48 @@ def save_sample(image, sample_path_i, filename, jpg_sample, write_info_files, wr
     else:
         image.save(f"{filename_i}.jpg", 'jpeg', quality=100, optimize=True)
     if write_info_files or write_sample_info_to_log_file:
-        # toggles differ for txt2img vs. img2img:
```
What is the purpose of this change?

hlky commented on scripts/webui.py:
```diff
@@ -704,20 +821,24 @@ def save_sample(image, sample_path_i, filename, jpg_sample, write_info_files, wr
             ignore_list = ["prompt", "target", "toggles", "ddim_eta", "batch_size"]
             rename_dict = {"ddim_steps": "steps", "n_iter": "number", "sampler_name": "sampler"} #changes the name of parameters to match with dynamic parameters
             sample_log_path = os.path.join(sample_path_i, "log.yaml")
-            log_dump = info_dict.get("prompt") # making sure the first item that is listed in the txt is the prompt text
```
Why was this changed?

hlky commented on scripts/webui.py:
```diff
@@ -704,20 +821,24 @@ def save_sample(image, sample_path_i, filename, jpg_sample, write_info_files, wr
 
-                if key == "cfg_scale": #adds zeros to to cfg_scale necessary for dynamic params
-                    value = str(value).zfill(2)
+                    if key == "cfg_scale": #adds zeros to to cfg_scale necessary for dynamic params
```
Avoid changing formatting. It makes the diff messy.

hlky commented on scripts/webui.py:
```diff
@@ -704,20 +821,24 @@ def save_sample(image, sample_path_i, filename, jpg_sample, write_info_files, wr
+                    if key == "cfg_scale": #adds zeros to to cfg_scale necessary for dynamic params
+                        value = str(value).zfill(2)
 
-                if found_key:
```
What is the purpose of this change?

hlky commented on scripts/webui.py:
```diff
@@ -704,20 +821,24 @@ def save_sample(image, sample_path_i, filename, jpg_sample, write_info_files, wr
+                    log_dump += f" {key} {value}"
+                log_dump = log_dump + " \n" #space at the end for dynamic params to accept the last param
 
-            log_dump = log_dump + " \n" #space at the end for dynamic params to accept the last param
```
Why was this moved an indent level?

hlky commented on scripts/webui.py:
```diff
@@ -1209,7 +1330,7 @@ def sample_iteration_callback(image_sample: torch.Tensor, iter_num: int):
                     gfpgan_filename = original_filename + '-gfpgan'
                     save_sample(gfpgan_image, sample_path_i, gfpgan_filename, jpg_sample, write_info_files, write_sample_info_to_log_file, prompt_matrix, init_img, uses_loopback, uses_random_seed_loopback, skip_save,
-skip_grid, sort_samples, sampler_name, ddim_eta, n_iter, batch_size, i, denoising_strength, resize_mode, skip_metadata=False)
+skip_grid, sort_samples, sampler_name, ddim_eta, n_iter, batch_size, i, denoising_strength, resize_mode, skip_metadata=False, custom_info_dict=None, custom_log_dump=None)
```
`custom_info_dict=None, custom_log_dump=None`
Not required.

hlky commented on scripts/webui.py:
```diff
@@ -1229,7 +1350,7 @@ def sample_iteration_callback(image_sample: torch.Tensor, iter_num: int):
                     ImageMetadata.set_on_image( esrgan_image, metadata )
                     save_sample(esrgan_image, sample_path_i, esrgan_filename, jpg_sample, write_info_files, write_sample_info_to_log_file, prompt_matrix, init_img, uses_loopback, uses_random_seed_loopback, skip_save,
-skip_grid, sort_samples, sampler_name, ddim_eta, n_iter, batch_size, i, denoising_strength, resize_mode, skip_metadata=False)
+skip_grid, sort_samples, sampler_name, ddim_eta, n_iter, batch_size, i, denoising_strength, resize_mode, skip_metadata=False, custom_info_dict=None, custom_log_dump=None)
```
`custom_info_dict=None, custom_log_dump=None`
Not required.

hlky commented on scripts/webui.py:
```diff
@@ -1251,7 +1372,7 @@ def sample_iteration_callback(image_sample: torch.Tensor, iter_num: int):
                     ImageMetadata.set_on_image(gfpgan_esrgan_image, metadata)
                     save_sample(gfpgan_esrgan_image, sample_path_i, gfpgan_esrgan_filename, jpg_sample, write_info_files, write_sample_info_to_log_file, prompt_matrix, init_img, uses_loopback, uses_random_seed_loopback,
-skip_save, skip_grid, sort_samples, sampler_name, ddim_eta, n_iter, batch_size, i, denoising_strength, resize_mode, skip_metadata=False)
+skip_save, skip_grid, sort_samples, sampler_name, ddim_eta, n_iter, batch_size, i, denoising_strength, resize_mode, skip_metadata=False, custom_info_dict=None, custom_log_dump=None)
```
`custom_info_dict=None, custom_log_dump=None`
Not required.

hlky commented on scripts/webui.py:
```diff
@@ -1269,7 +1390,7 @@ def sample_iteration_callback(image_sample: torch.Tensor, iter_num: int):
                 if not skip_save:
                     save_sample(image, sample_path_i, filename, jpg_sample, write_info_files, write_sample_info_to_log_file, prompt_matrix, init_img, uses_loopback, uses_random_seed_loopback, skip_save,
-skip_grid, sort_samples, sampler_name, ddim_eta, n_iter, batch_size, i, denoising_strength, resize_mode, False)
+skip_grid, sort_samples, sampler_name, ddim_eta, n_iter, batch_size, i, denoising_strength, resize_mode, False, custom_info_dict=None, custom_log_dump=None)
```
`custom_info_dict=None, custom_log_dump=None`
Not required.

hlky commented on scripts/webui.py:
```diff
@@ -1286,7 +1407,8 @@ def sample_iteration_callback(image_sample: torch.Tensor, iter_num: int):
                                     normalize_prompt_weights, use_GFPGAN, write_info_files, write_sample_info_to_log_file, prompt_matrix, init_img, uses_loopback, uses_random_seed_loopback, skip_save,
-                                    skip_grid, sort_samples, sampler_name, ddim_eta, n_iter, batch_size, i, denoising_strength, resize_mode, False)
+                                    skip_grid, sort_samples, sampler_name, ddim_eta, n_iter, batch_size, i, denoising_strength, resize_mode, False,
+                                    custom_info_dict=None, custom_log_dump=None)
```
`custom_info_dict=None, custom_log_dump=None`
Not required.

hlky commented on scripts/webui.py:
```diff
@@ -1834,6 +1956,1079 @@ def sample(init_data, x, conditioning, unconditional_conditioning, sampler_name,
     return output_images, seed, info, stats
 
 
+scn2img_cache = {
```
Can all of this be moved out of `webui.py`

hlky commented on scripts/webui.py:
```diff
@@ -2202,7 +3397,7 @@ def processLDSR(image):
                 if 1 not in imgproc_toggles:
                     output.append(image)
-                    save_sample(image, outpathDir, outFilename, False, None, None, None, None, None, False, None, None, None, None, None, None, None, None, None, False)
+                    save_sample(image, outpathDir, outFilename, False, None, None, None, None, None, False, None, None, None, None, None, None, None, None, None, False, None, None)
```
`custom_info_dict=None, custom_log_dump=None`
Not required.

hlky commented on scripts/webui.py:
```diff
@@ -2212,7 +3407,7 @@ def processLDSR(image):
                     outFilename = str(batchNumber)+'-'+'result'
                     output.append(image)
-                    save_sample(image, outpathDir, outFilename, False, None, None, None, None, None, False, None, None, None, None, None, None, None, None, None, False)
+                    save_sample(image, outpathDir, outFilename, False, None, None, None, None, None, False, None, None, None, None, None, None, None, None, None, False, None, None)
```
`custom_info_dict=None, custom_log_dump=None`
Not required.

hlky commented on scripts/webui.py:
```diff
@@ -2222,7 +3417,7 @@ def processLDSR(image):
                     outFilename = str(batchNumber)+'-'+'result'
                     output.append(image)
-                    save_sample(image, outpathDir, outFilename, False, None, None, None, None, None, False, None, None, None, None, None, None, None, None, None, False)
+                    save_sample(image, outpathDir, outFilename, False, None, None, None, None, None, False, None, None, None, None, None, None, None, None, None, False, None, None)
```
`custom_info_dict=None, custom_log_dump=None`
Not required.

hlky commented on scripts/webui.py:
```diff
@@ -2232,7 +3427,7 @@ def processLDSR(image):
                     outFilename = str(batchNumber)+'-'+'result'
                     output.append(image)
-                    save_sample(image, outpathDir, outFilename, False, None, None, None, None, None, False, None, None, None, None, None, None, None, None, None, False)
+                    save_sample(image, outpathDir, outFilename, False, None, None, None, None, None, False, None, None, None, None, None, None, None, None, None, False, None, None)
```
`custom_info_dict=None, custom_log_dump=None`
Not required.

hlky commented on scripts/webui.py:
```diff
@@ -2246,7 +3441,7 @@ def processLDSR(image):
                     output.append(image)
 
-                    save_sample(image, outpathDir, outFilename, None, None, None, None, None, None, False, None, None, None, None, None, None, None, None, None, False)
+                    save_sample(image, outpathDir, outFilename, None, None, None, None, None, None, False, None, None, None, None, None, None, None, None, None, False, None, None)
```
`custom_info_dict=None, custom_log_dump=None`
Not required.

hlky commented on scripts/webui.py:
```diff
@@ -2370,6 +3632,7 @@ def run_RealESRGAN(image, model_name: str):
     'variant_amount': 0.0,
     'variant_seed': '',
     'submit_on_enter': 'Yes',
+    'realesrgan_model_name': 'RealESRGAN_x4plus',
```
What is the purpose of this change?

hlky commented on scripts/webui.py:
```diff
@@ -2444,6 +3707,9 @@ def run_RealESRGAN(image, model_name: str):
     'height': 512,
     'width': 512,
     'fp': None,
+    'mask_blur_strength': 1,
```
What is the purpose of these changes?

xaedes replied:
For those two txt2img and img2img arguments the defaults where missing in txt2img_defaults and img2img_defaults.
I added the missing values because, otherwise their defaults are unknown, because set from the frontend only. 
For scn2img I use txt2img_defaults and the other one to populate not specified arguments in the txt2img, etc. calls.
I wanted the used defaults values to be the same as the user would expect from using the txt2img and img2img web frontend. When they are missing the calls would have use the default values in txt2img/img2img argument list, or fail and throw an exception when there is no default value specified.

xaedes commented:
@codedealer I added output info and batch view progress.

@hlky I moved the code, pinned the dependency versions and reverted changes to not scn2img-related code (save_sample). For the remaining changes, i.e. the added default values for img2img and txt2img, I think my https://github.com/sd-webui/stable-diffusion-webui/pull/1179#discussion_r979140541 explains the reasoning.

@Dekker3D  
I also added MiDaS based depth estimation and 3D image transformation. Not perfect, but actually usable.

Images can now be blended in other modes ([full list](https://pillow.readthedocs.io/en/stable/reference/ImageChops.html)) like `multiply` or `screen`. I used it mainly for better mask selection.

![Screenshot](https://i.imgur.com/KbRjty7.png)

codedealer reviewed (commented):
Couple points:
1. @hlky this currently pulls pickles from github for MiDaS into pytorch's cache. Should we pack them inside repo instead?
2. @xaedes During intermediate renders you call webui's save_sample with `skip_metadata` set to `False` hardcoded. In your example with the waterfall scene when saving empty masks this results in a warning thrown in console:
> Couldn't find metadata on image

It's not critical but can we do something about it?

hlky commented:
Note the approved files. I haven't tested this yet, but I have resolved the conflicts in `requirements.txt` and `environment.yaml`, there shouldn't be any issues with the accompanying data files.

@codedealer @xaedes what is the status of this?

xaedes commented:
@hlky Tested it and it does work with the updated environments.yaml. requirements.txt also looks good, but at my machine I can't test if it works with Docker.

@codedealer I will do something about the image metadata.
